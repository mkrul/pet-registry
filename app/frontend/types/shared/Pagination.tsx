export interface IPagination {
  count: number;
  items: number;
  pages?: number;
  page: number;
}
export interface IPaginationQuery {
  page?: number;
  items?: number;
  query?: string;
}
