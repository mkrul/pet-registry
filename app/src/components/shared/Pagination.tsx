import React from "react";

interface PaginationProps {
  currentPage: number;
  totalPages: number;
  onPageChange: (page: number) => void;
}

const Pagination: React.FC<PaginationProps> = ({ currentPage, totalPages, onPageChange }) => {
  return (
    <div className="flex justify-between mt-4">
      <button
        onClick={() => onPageChange(currentPage - 1)}
        disabled={currentPage === 1}
        className={`px-4 py-2 rounded ${
          currentPage === 1
            ? "bg-gray-300 cursor-not-allowed"
            : "bg-blue-500 text-white hover:bg-blue-600"
        }`}
      >
        Previous
      </button>
      <span>
        Page {currentPage} of {totalPages}
      </span>
      <button
        onClick={() => onPageChange(currentPage + 1)}
        disabled={currentPage === totalPages}
        className={`px-4 py-2 rounded ${
          currentPage === totalPages
            ? "bg-gray-300 cursor-not-allowed"
            : "bg-blue-500 text-white hover:bg-blue-600"
        }`}
      >
        Next
      </button>
    </div>
  );
};

export default Pagination;
