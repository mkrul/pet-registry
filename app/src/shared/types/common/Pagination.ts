export interface PaginationProps {
  count: number;
  page: number;
  items: number;
  pages: number;
  per_page: number;
}
export interface PaginationPropsQuery {
  page: number;
  items: number;
  query?: string;
  species?: string;
  color?: string;
  gender?: string;
  sort?: string;
  country?: string;
  state?: string;
  area?: string;
  breed?: string;
  archived?: boolean;
}

export interface PaginationPagesProps {
  currentPage: number;
  totalPages: number;
  onPageChange: (page: number) => void;
}
