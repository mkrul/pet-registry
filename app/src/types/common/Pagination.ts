export interface PaginationProps {
  count: number;
  page: number;
  items: number;
  pages: number;
}

export interface PaginationPagesProps {
  currentPage: number;
  totalPages: number;
  onPageChange: (page: number) => void;
}
