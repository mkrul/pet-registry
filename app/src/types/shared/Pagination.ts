export interface IPaginationQuery {
  page: number;
  items: number;
  query?: string;
  species?: string;
  color?: string;
  gender?: string;
  sort?: string;
}

export interface IPagination {
  current_page: number;
  total_pages: number;
  total_count: number;
  per_page: number;
}
