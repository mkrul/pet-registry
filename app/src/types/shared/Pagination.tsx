export interface IPagination {
  count: number;
  items: number;
  pages: number;
  page: number;
}
export interface IPaginationQuery {
  page: number;
  items: number;
  query?: string;
  species?: string;
  color?: string;
  gender?: string;
  sort?: string;
  country?: string;
  state?: string;
}
