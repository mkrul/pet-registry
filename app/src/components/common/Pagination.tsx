import React from "react";
import { PaginationPagesProps } from "../../types/common/Pagination";

const Pagination: React.FC<PaginationPagesProps> = ({ currentPage, totalPages, onPageChange }) => {
  return (
    <div className="flex justify-between mt-4">
      <button
        onClick={() => onPageChange(currentPage - 1)}
        disabled={currentPage === 1}
        className={`px-4 py-2 rounded ${
          currentPage === 1
            ? "bg-gray-300 cursor-not-allowed"
            : "bg-blue-500 text-white hover:bg-blue-600"
        }`}
      >
        Previous
      </button>
      <span>
        Page {currentPage} of {totalPages}
      </span>
      <button
        onClick={() => onPageChange(currentPage + 1)}
        disabled={currentPage === totalPages}
        className={`px-4 py-2 rounded ${
          currentPage === totalPages
            ? "bg-gray-300 cursor-not-allowed"
            : "bg-blue-500 text-white hover:bg-blue-600"
        }`}
      >
        Next
      </button>
    </div>
  );
};

export default Pagination;
