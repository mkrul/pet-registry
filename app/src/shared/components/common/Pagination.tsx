import React from "react";
import { PaginationPagesProps } from "../../../../shared/types/common/Pagination";

const Pagination: React.FC<PaginationPagesProps> = ({ currentPage, totalPages, onPageChange }) => {
  return (
    <div className="flex justify-center items-center mt-4">
      <div className="flex items-center gap-5">
        <button
          onClick={() => onPageChange(currentPage - 1)}
          disabled={currentPage === 1}
          className={`px-4 py-2 rounded ${
            currentPage === 1
              ? "bg-gray-300 cursor-not-allowed"
              : "bg-blue-500 text-white hover:bg-blue-600"
          }`}
        >
          Previous
        </button>
        <span className="text-gray-700 min-w-[100px] text-center">
          Page {currentPage} of {totalPages}
        </span>
        <button
          onClick={() => onPageChange(currentPage + 1)}
          disabled={currentPage === totalPages}
          className={`px-4 py-2 rounded ${
            currentPage === totalPages
              ? "bg-gray-300 cursor-not-allowed"
              : "bg-blue-500 text-white hover:bg-blue-600"
          }`}
        >
          Next
        </button>
      </div>
    </div>
  );
};

export default Pagination;
