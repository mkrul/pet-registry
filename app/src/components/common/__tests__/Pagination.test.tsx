import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/vitest";
import userEvent from "@testing-library/user-event";
import Pagination from "../Pagination";

describe("Pagination", () => {
  const mockOnPageChange = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
  });

  const renderPagination = (props = {}) => {
    return render(
      <Pagination currentPage={1} totalPages={10} onPageChange={mockOnPageChange} {...props} />
    );
  };

  describe("Rendering", () => {
    it("renders pagination controls", () => {
      renderPagination();
      expect(screen.getByText("Previous")).toBeDefined();
      expect(screen.getByText("Next")).toBeDefined();
    });

    it("displays current page and total pages", () => {
      renderPagination({ currentPage: 3, totalPages: 5 });
      expect(screen.getByText("Page 3 of 5")).toBeDefined();
    });

    it("disables Previous button on first page", () => {
      renderPagination({ currentPage: 1 });
      const prevButton = screen.getByText("Previous");
      expect(prevButton).toBeDisabled();
      expect(prevButton).toHaveClass("bg-gray-300", "cursor-not-allowed");
    });

    it("disables Next button on last page", () => {
      renderPagination({ currentPage: 10, totalPages: 10 });
      const nextButton = screen.getByText("Next");
      expect(nextButton).toBeDisabled();
      expect(nextButton).toHaveClass("bg-gray-300", "cursor-not-allowed");
    });

    it("enables both buttons when on middle page", () => {
      renderPagination({ currentPage: 5, totalPages: 10 });
      const prevButton = screen.getByText("Previous");
      const nextButton = screen.getByText("Next");
      expect(prevButton).not.toBeDisabled();
      expect(nextButton).not.toBeDisabled();
      expect(prevButton).toHaveClass("bg-blue-500", "hover:bg-blue-600");
      expect(nextButton).toHaveClass("bg-blue-500", "hover:bg-blue-600");
    });
  });

  describe("Interactions", () => {
    it("calls onPageChange with previous page number when Previous is clicked", async () => {
      renderPagination({ currentPage: 5 });
      const prevButton = screen.getByText("Previous");
      await userEvent.click(prevButton);
      expect(mockOnPageChange).toHaveBeenCalledTimes(1);
      expect(mockOnPageChange).toHaveBeenCalledWith(4);
    });

    it("calls onPageChange with next page number when Next is clicked", async () => {
      renderPagination({ currentPage: 5 });
      const nextButton = screen.getByText("Next");
      await userEvent.click(nextButton);
      expect(mockOnPageChange).toHaveBeenCalledTimes(1);
      expect(mockOnPageChange).toHaveBeenCalledWith(6);
    });

    it("does not trigger click events on disabled Previous button", async () => {
      renderPagination({ currentPage: 1 });
      const prevButton = screen.getByText("Previous");
      await userEvent.click(prevButton);
      expect(mockOnPageChange).not.toHaveBeenCalled();
    });

    it("does not trigger click events on disabled Next button", async () => {
      renderPagination({ currentPage: 10, totalPages: 10 });
      const nextButton = screen.getByText("Next");
      await userEvent.click(nextButton);
      expect(mockOnPageChange).not.toHaveBeenCalled();
    });
  });

  describe("Accessibility", () => {
    it("disables buttons appropriately for screen readers", () => {
      renderPagination({ currentPage: 1 });
      expect(screen.getByText("Previous")).toHaveAttribute("disabled");
      expect(screen.getByText("Next")).not.toHaveAttribute("disabled");
    });
  });

  describe("Snapshots", () => {
    it.each([
      { currentPage: 1, totalPages: 10, desc: "first page" },
      { currentPage: 5, totalPages: 10, desc: "middle page" },
      { currentPage: 10, totalPages: 10, desc: "last page" }
    ])("matches snapshot for $desc", ({ currentPage, totalPages }) => {
      const { container } = renderPagination({ currentPage, totalPages });
      expect(container).toMatchSnapshot();
    });
  });
});
