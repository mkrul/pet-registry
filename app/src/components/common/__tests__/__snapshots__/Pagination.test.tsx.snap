// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination > Snapshots > matches snapshot for 'first page' 1`] = `
<div>
  <div
    class="flex justify-between mt-4"
  >
    <button
      class="px-4 py-2 rounded bg-gray-300 cursor-not-allowed"
      disabled=""
    >
      Previous
    </button>
    <span>
      Page 
      1
       of 
      10
    </span>
    <button
      class="px-4 py-2 rounded bg-blue-500 text-white hover:bg-blue-600"
    >
      Next
    </button>
  </div>
</div>
`;

exports[`Pagination > Snapshots > matches snapshot for 'last page' 1`] = `
<div>
  <div
    class="flex justify-between mt-4"
  >
    <button
      class="px-4 py-2 rounded bg-blue-500 text-white hover:bg-blue-600"
    >
      Previous
    </button>
    <span>
      Page 
      10
       of 
      10
    </span>
    <button
      class="px-4 py-2 rounded bg-gray-300 cursor-not-allowed"
      disabled=""
    >
      Next
    </button>
  </div>
</div>
`;

exports[`Pagination > Snapshots > matches snapshot for 'middle page' 1`] = `
<div>
  <div
    class="flex justify-between mt-4"
  >
    <button
      class="px-4 py-2 rounded bg-blue-500 text-white hover:bg-blue-600"
    >
      Previous
    </button>
    <span>
      Page 
      5
       of 
      10
    </span>
    <button
      class="px-4 py-2 rounded bg-blue-500 text-white hover:bg-blue-600"
    >
      Next
    </button>
  </div>
</div>
`;
